### PR TITLE
Fix bug with MateDistantReadFilter not being configurable

### LINKED ISSUE #7696: Bug in HaplotypeCaller: parameter mate-too-distant-length is readonly
### Affected tool(s) or class(es)
HaplotypeCaller

### Affected version(s)
- Latest public release version [4.2.5.0]

#### Steps to reproduce
Specify `--mate-too-distant-length` parameter.

#### Expected behavior
Use specified value to filter paired-end reads.

#### Actual behavior
Java exception:

```
Using GATK jar /home/rwilton/tools/gatk-4.2.5.0/gatk-package-4.2.5.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /home/rwilton/tools/gatk-4.2.5.0/gatk-package-4.2.5.0-local.jar HaplotypeCaller
 --reference /datascope/rwilton/scratch/GATK/GRCh38/chr14.fna
 --intervals chr14
 --emit-ref-confidence GVCF
 --sample-name HG002
 --smith-waterman FASTEST_AVAILABLE
 --native-pair-hmm-threads 48
 --read-filter MateDistantReadFilter
 --mate-too-distant-length 1500
 --minimum-mapping-quality 10
 --mapping-quality-threshold-for-genotyping 10
 --input full.chr14.bam
 --output vcf1/full.chr14.g.vcf.gz
org.broadinstitute.barclay.argparser.CommandLineException$ShouldNeverReachHereException: Couldn't set field value for mateTooDistantLength in org.broadinstitute.hellbender.engine.filters.MateDistantReadFilter@55fdf7f9 with value 1500.
        at org.broadinstitute.barclay.argparser.NamedArgumentDefinition.setArgumentValue(NamedArgumentDefinition.java:680)
        at org.broadinstitute.barclay.argparser.NamedArgumentDefinition.setScalarValue(NamedArgumentDefinition.java:380)
        at org.broadinstitute.barclay.argparser.NamedArgumentDefinition.setArgumentValues(NamedArgumentDefinition.java:293)
        at org.broadinstitute.barclay.argparser.CommandLineArgumentParser.propagateParsedValues(CommandLineArgumentParser.java:490)
        at org.broadinstitute.barclay.argparser.CommandLineArgumentParser.parseArguments(CommandLineArgumentParser.java:170)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.parseArgs(CommandLineProgram.java:233)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:207)
        at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
        at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
        at org.broadinstitute.hellbender.Main.main(Main.java:289)
Caused by: java.lang.IllegalAccessException: Can not set static final int field org.broadinstitute.hellbender.engine.filters.MateDistantReadFilter.mateTooDistantLength to java.lang.Integer
        at sun.reflect.UnsafeFieldAccessorImpl.throwFinalFieldIllegalAccessException(UnsafeFieldAccessorImpl.java:76)
        at sun.reflect.UnsafeFieldAccessorImpl.throwFinalFieldIllegalAccessException(UnsafeFieldAccessorImpl.java:80)
        at sun.reflect.UnsafeQualifiedStaticIntegerFieldAccessorImpl.set(UnsafeQualifiedStaticIntegerFieldAccessorImpl.java:77)
        at java.lang.reflect.Field.set(Field.java:764)
        at org.broadinstitute.barclay.argparser.NamedArgumentDefinition.setArgumentValue(NamedArgumentDefinition.java:677)
        ... 9 more
HaplotypeCaller ends (3)
```

### PR DESCRIPTION
**Pull request by jamesemery**
Fixes #7696

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/filters/MateDistantReadFilter.java [hunk 1]
```diff
@@ -14,10 +14,18 @@
...
     private static final long serialVersionUID = 1L;
 
     @Argument(fullName = ReadFilterArgumentDefinitions.MATE_TOO_DISTANT_LENGTH, doc="Minimum start location difference at which mapped mates are considered distant", optional=true)
-    public static final int mateTooDistantLength = 1000;
+    public int mateTooDistantLength = DEFAULT_MATE_TOO_DISTANT_THRESHOLD;
+
+    public MateDistantReadFilter() {
+    }
+
+    public MateDistantReadFilter( final int minMappingQualityScore ) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/filters/MateDistantReadFilter.java:26]: This is not a mapping quality score.

**Review by lbergelson — CHANGES_REQUESTED** [review]
👎

**Comment by lbergelson** [comment]
In all seriousness though, fix the typo and merge :)

**Review by lbergelson — APPROVED** [review]
👍

**Comment by RWilton** [comment]
Sorry, but this bug still isn't fixed as of v4.2.6.1.  Reproduce as follows:

```
  --read-filter MateDistantReadFilter
  --mate-too-distant-length 1500
```

Instead of a run-time exception (as in v4.2.5.0), HaplotypeCaller simply produces no variant calls at all.

Expected behavior would be to exclude paired-end mappings whose TLEN exceeds the parameterized value.  Perhaps there is an implementation bug, unrelated to the original problem, that contains faulty logic for doing this.

Thanks...

**Comment by jamesemery** [comment]
Hi @RWilton i'm sorry to hear that. I suspect its unrelated given how simple this PR is but its quite possible that filter has been broken since evidently nobody was able to use and adjust it for a long time. This is the logic here that the filter uses:
`return read.isPaired() && !read.isUnmapped() && !read.mateIsUnmapped() &&
                (Math.abs(read.getStart() - read.getMateStart()) >= mateTooDistantLength || !read.getContig().equals(read.getMateContig()));`

That logic is correct for what the filter is doing. It should be noted that the filter does the opposite of what you expect it to (since its intended for our SV pipeline) in that it filters out all reads that are NOT having distant mates. This means if you try to run HaplotypeCaller with this setting you will be throwing away every read pair EXCEPT the distant ones which results in mostly no reads. We should perhaps rename this filter to be a little less confusing.

**Comment by RWilton** [comment]
Well, that explains that, sort of.

The code snippet you're providing looks like it ought to do what you say it does (i.e., the mates have to be paired, not unmapped, mapped to the same contig, and have a difference in their start positions that is at least `mateTooDistantLength`).  

But there are two problems with this:

1) This filter's behavior is unexpected wrt HaplotypeCaller.  It seems to me that an inclusive filter (i.e., process only paired-end mappings whose TLEN falls within a specified range) would be more usable.  That would imply a filter implementation that accepts a pair of integers, but the expected behavior would be more obvious and in line with GATK's other range-limited parameterizations (e.g., `MappingQualityReadFilter` comes immediately to mind).

2) I can't tell from where I sit, but the code snippet looks correct only if `getStart()` and `getMateStart()` return a zero-based start position of each mate relative to the start of the strand to which the mate is mapped.  If the code is just computing the difference between POS for the mates, the computation is incorrect for forward + reverse-complement (Illumina-style) pairs.  In addition, computing TLEN requires not only that you consider the orientation of the individual mate mappings, but also that you make an arbitrary decision about how to handle soft-clipped reads.

I hate to say this, but I think this parameter needs some attention.  Its potential utility with HaplotypeCaller seems evident to me (i.e., it would be good to be able to exclude outliers with unreasonable TLENs) but its implementation and frugal documentation make it unusable in practice.

**Comment by jamesemery** [comment]
@RWilton I agree that this is a confusing filter and maybe should be renamed... As I have said this is a specific utility is for SVs where they might only want to look at long range read pairs not HaplotypeCaller. In that context I would say that the nuance about TLEN vs left aligned start position is a relatively small one. I agree with you that it would be useful to have some functionality for filtering out highly distant read pairs form our short variant calling as an option. I would have to think on whether its better to refactor this filter to be more general or to make a second filter that is the reverse of this one. 

@RWilton can you make an issue ticket and describe what you would want out of such a filter for HaplotypeCaller? I would put some thought into the case where the TLEN field has not been populated since it often isn't (especially since the cases where TLEN might not get computed overlaps with discordant/distant mates which are part of the target for such a filter).

**Comment by RWilton** [comment]
Since the intent and implementation of this filter are not essential to my work -- I can simply filter pairs on TLEN prior to variant calling -- perhaps it's best to just let this one be.

Thanks for the discussion and for helping me think things through.

**Pull request merged**